### PR TITLE
Add Dockerfile

### PR DESCRIPTION
**Pull request by jferris**
This adds a Dockerfile that can serve as a reference point for
Suspenders-like applications which will be deployed to OCI platforms
like Docker, Kubernetes, and ECS.

### CODE REVIEW

**Comment by mike-burns** [comment]
Should this be in a generator? I don't actually understand what this file is doing.

**Comment by jferris** [comment]
> Should this be in a generator?

I'm not sure where it should be, but I'm guessing that if it were in a generator, nobody would ever notice it again.

> I don't actually understand what this file is doing.

Having this file present means that running `docker build` or pushing to a pipeline that supports Docker will result in having a Docker image. It's similar to a Heroku buildpack, which tells Heroku how to make your application into a slug, whereas this tells a Docker-compatible builder how to make your application into a container image.

I think my preference would be to have it new repositories, similar to having `.tool-versions`. If you don't use ASDF or anything, that file does nothing. A `Dockerfile` is the same.

More than anything, I want to have a place where we document and collaborate on a good stock `Dockerfile` for Suspenders projects, and Suspenders seems like a good place to start.

**Comment by mike-burns** [comment]
Oooh. I read the PR description as if you were adding this as a sample Dockerfile for projects, which is confusing because that's the point of Suspenders.

But you're adding this Dockerfile so you can use Docker with Suspenders? That makes much more sense.

**Comment by mike-burns** [comment]
Anyway, my feelings on Docker is: does this have to live in the repo? Can we have a separate repo of Dockerfiles for the people who do use Docker?

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  && bundle config --global frozen 1 \
+  && bundle config --global without "development test" \
+  && bundle install -j4 --retry 3 \
+  && rm -rf /usr/local/bundle/cache/*.gem \
+  && find /usr/local/bundle/gems/ -name "*.c" -delete \
+  && find /usr/local/bundle/gems/ -name "*.o" -delete
+
+COPY . /app
+
+# The SECRET_KEY_BASE here isn't used. Precomiling assets doesn't use your
```
Comment (johnpaulashenfelter) [inline Dockerfile:33]: Type: "... Precompiling"

Comment (jferris) [reply Dockerfile]: Fixed in 1876ed3.

**Comment by thiagoa** [comment]
I like this idea. As @mike-burns noted, this could become part of some deployment generator: we'd move the Heroku logic into a separate generator and the `Dockerfile` would be part of another generator, yet to be named. That would allow us to choose Heroku or something else if that's what we want.

However, as the `Dockerfile` is at the root of `suspenders`, how would it get copied to actual applications? Should we place this file in `templates` and add a build instruction to copy it over to the root of the generated application? The template could have an `.erb` extension that would allow us to inject the Ruby version used by the client (probably the `RUBY_VERSION` constant), but then there's the question: if we inject the Ruby version, would a Docker image always exist for it? Currently, the Ruby version is hardcoded in Suspenders, but I'm working on making it flexible: https://github.com/thoughtbot/suspenders/issues/1073 

Since we don't have a defined idea of what generator we will use for the `Dockerfile`, I think we could simply go with this older-school solution. Being in a generator would still allow us to run that generator by default when running `bin/suspenders`.

**Comment by stevepolitodesign** [comment]
Would [MRSK](https://github.com/mrsked/mrsk) solve for this?

**Comment by jferris** [comment]
> Would [MRSK](https://github.com/mrsked/mrsk) solve for this?

No, I don't believe so. I haven't used MRSK but it looks like an orchestration platform for Docker containers. You'd still need a Dockerfile (or something else) to containerize the application.

**Comment by stevepolitodesign** [comment]
Actually, it looks like Rails [will ship with a Dockerfile](https://github.com/rails/rails/blob/main/railties/lib/rails/generators/rails/app/templates/Dockerfile.tt) in the next release?

**Comment by jferris** [comment]
If that's the case, will Suspenders automatically get that when we update it for that version of Rails?

**Comment by mike-burns** [comment]
Yup!

**Comment by mike-burns** [comment]
That said, if we want a different one, we should add a `production:deploy` generator that generates the Dockerfile we'd prefer.

**Comment by jferris** [comment]
I looked at the template for the Dockerfile in Rails and it seems like a good starting place for most applications. We might have to make changes for specific deployment platforms, but it's enough to get us going.

I'm going to close this pull request.

**Pull request closed without merging**